src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs

```diff
@@ -157,12 +157,17 @@ public bool HasPendingSymbolAnalysis(ISymbol symbol)
             {
                 if (pendingEntities.TryGetValue(analysisEntity, out var state))
                 {
+                    if (state != null && state.StateKind == StateKind.InProcess)
```

Comment:
>state.StateKind == StateKind.InProcess [](http://example.com/codeflow?start=41&length=38)

When will this be true? Doesn't the change to `AnalyzerExecutor.TryExecuteCompilationActions()` ensure that only the current thread is processing `analysisEntity`?

Reply:
There is still a small window between when we reset the analyzer state to ready after executing the actions AND before attempt to marking completion when another thread might invoke `TryExecuteCompilationActions` and get hold of the analyzer state and start processing it. The second thread will not make any callbacks as prior thread made all callbacks, but with this change we now ensure that the second thread marks completion instead of the first one doing it while second thread is executing and has hold of the state object.